File: assignment2/rq.py
import numpy as np

def rq(a):
    """
    Perform RQ factorization on matrix 'a' such that a = RQ,
    where R is an upper triangular matrix and Q is a unitary matrix.
    
    Parameters:
        a (np.ndarray): The input matrix to be factorized.
    
    Returns:
        r (np.ndarray): The upper triangular matrix R.
        q (np.ndarray): The unitary matrix Q.
    """
    m, n = a.shape
    e = np.eye(m)
    p = e[:, ::-1]  # reverse the order of columns to create permutation matrix
    q0, r0 = np.linalg.qr((p @ a[:, :m].T) @ p)

    r = (p @ r0.T) @ p
    q = (p @ q0.T) @ p

    # Ensure R has positive diagonal elements by adjusting with a fix matrix
    fix = np.diag(np.sign(np.diag(r)))
    r = r @ fix
    q = fix @ q

    # If a is not square, augment Q
    if n > m:
        q = np.hstack((q, np.linalg.inv(r) @ a[:, m:n]))

    return r, q

File: assignment4/A4code/rq.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% EEN020 - Computer Vision 2021
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% function: rq
%
% RQ factorization. Factorizes a such a=rq where r upper tri. and q unit matrix
% If a is not square, then q is equal q=[q1 q2] where q1 is unit matrix
% 
%   inputs:    a: (M, N) array, M<=N
%
%   outputs:   r: (M, M) array
%                 upper triangular matrix
%              q: (M, N) array
%                 unit matrix; if a is not square, then q=[q1 q2]
%                 where q1 is unit matrix
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

function [r,q]=rq(a)
[m,n]=size(a);
e=eye(m);
p=e(:,[m:-1:1]);
[q0,r0]=qr(p*a(:,1:m)'*p);

r=p*r0'*p;
q=p*q0'*p;

fix=diag(sign(diag(r)));
r=r*fix;
q=fix*q;

if n>m
  q=[q, inv(r)*a(:,m+1:n)];
end
